src/lib.rs

```diff
@@ -525,6 +547,8 @@ pub enum Error {
     ImpossibleSatisfaction,
     /// Bare descriptors don't have any addresses
     BareDescriptorAddr,
+    /// PubKey invalid under current context
+    PubKeyCtxError(String, String),
```

Comment:
Why we have to use `String` here? Because we do not have a type enumerating all possible public keys in `rust-bitcoin`? But we can still use a specific context type with `dyn ScriptContext` I assume...

Another thing: I prefer to use named associated fields when it is not obvious which type contains which informtion:

```suggestion
    PubKeyCtxError { pubkey: String, context: String },
```

Reply:
> It is also not clear why we can't just require ScriptContext: Display, implement Display for specific context types and get rid from this method here

Replying to the above comment first because GitHub is not letting me reply there. :shrug: 
`ScriptContext` is never implemented on any object that can be created. Therefore, this is no `Self` to impl `fmt::Display` on. In our usage, `ScriptContext` is just used as a generic parameter with no objects ever created. Similar to `secp256k1::Signing` and `secp256k1::Verification`. 

```
// empty enum. No object can be created for this
pub enum Legacy {}
```

> Why we have to use String here? Because we do not have a type enumerating all possible public keys in rust-bitcoin? But we can still use a specific context type with dyn ScriptContext I assume...

The first pubkey string is my laziness. I don't see any reason why I can't return `KeyParseError`. The second one has answers in this comment and other comments.

Reply:
Then, the question is why we use non-instantiatable enum `pub enum Legacy {}` instead of unit struct `pub struct Legacy;`? In the second case the form will remain the same (`MiniscriptKey<..., Tap>`) and we will be able to add `where Self: ` conditions

Reply:
I think Andrew liked this cannot be instantiated because it does not make sense to have a context object. I don't really mind changing it, but @apoelstra has some opinions on this.